[ZEPPELIN-732] Helium Application

Leemoonsoo opened the pull request:
### What is this PR for?
This PR implements pluggable Application Framework.

Here's summary of the change.

**Applicaiton**

Base class of user application. User application should extends `org.apache.zeppelin.helium.Application`.

It looks like

```java
public abstract class Application {

  /**
   * When application loaded
   * @params context Application context
   */
  public Application(ApplicationContext context) {
    this.context = context;
  }

  /**
   * This method can be invoked multiple times before unload(),
   * Invoked just after application loaded or when paragraph output is updated
   * @param args Required resources from paragraph
   */
  public abstract void run(ResourceSet args)
      throws ApplicationException, IOException;


  /**
   * this method is invoked just before application is removed
   */
  public abstract void unload() throws ApplicationException;
}
```


**HeliumPackage**

`org.apache.zeppelin.helium.HeliumPackage` keeps information of application package.
Package informations includes

```
{
  type : "APPLICATION",                            // type of this package
  name : "zeppelin.app",                           // application name. [organizationName].[appName] format
  description : "App description",                 // description
  artifact : "groupId:artifactId:version",         // artifact name for maven repository
  className : "org.apache.zeppelin.helium.MyApp",  // application class that extends org.apache.zeppelin.helium.Application
  resources : [[]],                                // required resource
  icon : '<i class="fa fa-clock-o"></i>'           // icon
}
```

`resources` field defines what kind of data this application requires, from ResourcePool.
"resourceName" and ":className" works in the array.
inner array combines each resourceName, :className with 'AND'
outer array combines inner array with 'OR'.

For example,

```
resources : [
    [ "name1", ":java.util.Date"],
    [ "name1", "name2"]
]
```

Then Zeppelin searches ResourcePool, first for resource name "name1" and resource type "java.util.Date" from resourcePool then resource name "name1" and "name2".

Once resources are found, they'll be passed to `Application.run()` method

**Package Registry**

`org.apache.zeppelin.helium.HeliumRegistry` provides list of available HeliumPackages.
Currently `HeliumLocalRegistry` is implemented and it provides list by reading HeliumPackage json files under ./helium directory in the file system. Later there will be some class like `HeliumOnlineRegistry'  which reads available package from a community managed online central registry.

**Development mode**
`org.apache.zeppelin.interpreter.dev.*` package is provided for development support.
Developer can run their app inside of their IDE, that connects to running Zeppelin instance and display output.

**Application Suggestion**
Once paragraph becomes FINISHED status after run, Zeppelin searches resources from ResourcesPool that belongs to current paragraph. And compare all HeliumPackage if these resources satisfies their requirements ('resources' field). If there're available app, Helium button will be displayed.

![image](https://cloud.githubusercontent.com/assets/1540981/14518452/d8f85d5a-020f-11e6-8313-00b37eb4f077.png)

User can see and select available application for this paragraph.
![image](https://cloud.githubusercontent.com/assets/1540981/14518465/ee607fce-020f-11e6-8ad4-fe98723c7c94.png)

**Application selection**
After application is loaded, application icon will be displayed. So user can select and switch between apps.

![image](https://cloud.githubusercontent.com/assets/1540981/14518547/6f3fd766-0210-11e6-9a08-0dc2ba6d15e3.png)

if built-in visualization is available, application icon will be displayed next to built-in visualizations
![image](https://cloud.githubusercontent.com/assets/1540981/14518500/27505746-0210-11e6-9738-fb514376b906.png)



**Experimental**

While this is new feature and API and specification may change, i'd suggest mark this feature as a experimental for a while. 


### What type of PR is it?
Feature

### Todos
* [x] - Helium Application framework
* [x] - App launcher
* [x] - App display and selectors
* [x] - Package registry
* [x] - Development mode
* [x] - Make CI green
* [x] - Improve comment in source codes
* [x] - Documentation
* [x] - Examples
* [x] - Mark experimental
* [ ] - Review

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-732

### How should this be tested?
Will be updated with examples

### Screenshots (if appropriate)
![helium](https://cloud.githubusercontent.com/assets/1540981/14518685/49c2a788-0211-11e6-81c8-6cae48bf8a48.gif)


### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? yes

jongyoul commented:
@Leemoonsoo I've thought Helium is a kind of backend system. But as I saw your screenshot, it looks like a frontend system. Do you have any change? or does Helium support both of them?

Leemoonsoo commented:
@jongyoul Thanks for the interest.

Helium just runs jar library (which extends `Application`). Which may include some output and the output is displayed in Zeppelin notebook. And the output is always handled by Angular display system. So displayed output can interact with the code that runs on backend.

Does this approach makes sense? what do you think?

Leemoonsoo commented:
It's ready for review. This PR provides experimental Java API to add pluggable components.
Which is still i think pretty low level.

Adding high level api (e.g. [ZEPPELIN-1003](https://issues.apache.org/jira/browse/ZEPPELIN-1003)) and providing API for other languages (e.g. scala, python, javascript, etc) will be the future work.

bzz commented:
Great stuff!! It's going to take some time to review.

@Leemoonsoo  @minahlee Please help me understand, do you think this should be included in 0.6.0 release? As I do not see [ZEPPELIN-732](https://issues.apache.org/jira/browse/ZEPPELIN-732) under [ZEPPELIN-889](https://issues.apache.org/jira/browse/ZEPPELIN-889)

bzz commented on zeppelin-examples/zeppelin-example-horizontalbar/zeppelin-example-horizontalbar.json:
```diff
@@ -0,0 +1,25 @@
+ * limitations under the License.
+ */
+{
+  type : "APPLICATION",
```
Accoding to [rfc](http://tools.ietf.org/html/rfc7159) to be a valid JSON keys must be a strings, so are " needed

bzz commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreter.java:
```diff
@@ -76,17 +83,46 @@ public RemoteInterpreter(Properties property,
-      Map<String, String> env,
-      int connectTimeout,
-      RemoteInterpreterProcessListener remoteInterpreterProcessListener) {
+                           String noteId,
```
Is there any particular reason for changing formatting style here and above? I belive first one was more concise and used in Zeppelin consistently

Leemoonsoo replied:
I don't have preference of code formatting, but my IDE seems it does. 
I'll restore the style.

bzz commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NoteEventListener.java:
```diff
@@ -0,0 +1,12 @@
+package org.apache.zeppelin.notebook;
```
I belive that apache license header going to be needed here

bzz commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookEventListener.java:
```diff
@@ -0,0 +1,13 @@
+package org.apache.zeppelin.notebook;
```
I belive that apache license header going to be needed here

bzz commented:
@khalidhuseynov Just FYI, this PR adds 4 new messages websocket protocol:

```
APP_APPEND_OUTPUT,      // [s-c] append output
APP_UPDATE_OUTPUT,      // [s-c] update (replace) output
APP_LOAD,                           // [s-c] on app load
APP_STATUS_CHANGE,      // [s-c] on app status change
```

Leemoonsoo commented:
@bzz Pluggable visualization is one of [roadmap of 0.6 release](https://cwiki.apache.org/confluence/display/ZEPPELIN/Zeppelin+Roadmap), and i think this PR provides a way to enable it.

So if we can merge it into 0.6, it'll be better. but this feature will not be blocker for 0.6.0 release.

bzz commented:
Looks great, do you plan to rebase on master as well?

I think there was also @anish18sun using it under [ZEPPELIN-684](https://issues.apache.org/jira/browse/ZEPPELIN-684) so \cc @anish18sun for feedback as well

anish18sun commented:
read all the code changes, I think the usage of print() and println() around context().out.write() in Application.java is really intuitive for users!

Leemoonsoo commented:
Addressed comment from @bzz. @anish18sun Thanks for taking a look.

Now CI is green and i'll keep this PR open in this weekend for further reviews. if there're no further discussions, then i'll merge it into the master.

AhyoungRyu commented on docs/development/writingzeppelinapplication.md:
```diff
@@ -0,0 +1,185 @@
+-->
+{% include JB/setup %}
+
+### What is Zeppelin Application (Experimental)
```
@Leemoonsoo How about changing this  title to "What is Apache Zeppelin Application? (Experimental)" with using `<h1>` instead of `<h3>`?

AhyoungRyu commented on docs/development/writingzeppelinapplication.md:
```diff
@@ -0,0 +1,185 @@
+
+### What is Zeppelin Application (Experimental)
+
+Zeppelin Application is a package that runs on Interpreter process and display it's output inside of the notebook. While application runs on Interpreter process, it's able to access resources provided by Interpreter through ResourcePool. Output is always rendered by AngularDisplaySystem. Therefore application provides all the possiblity of making interactive graphical application that uses data a ...
```
- Since it's a first sentence of this docs, so i would suggest to start with "Apache Zeppelin Application ~" instread of "Zeppelin Application ~"
 - `display` -> `displays`
 - `possiblity` -> `possibilities`

AhyoungRyu commented on docs/development/writingzeppelinapplication.md:
```diff
@@ -0,0 +1,185 @@
+
+
+
+### Package flie
```
`flie -> `file`

Leemoonsoo commented:
Thanks @AhyoungRyu for the review. Comments are addressed.

AhyoungRyu commented:
@Leemoonsoo Thanks. Awesome work indeed!

bigloong commented:
Hi @Leemoonsoo , I have a question about Helium Application.

How to install the application for Zeppelin? And I do not find any application in Zeppelin.
The application format is Jar Format. When we want to deploy the application in production. we should put the jar package into which folder.

Thanks

Leemoonsoo commented:
jar package needs to be in the classpath of the interpreter process. I think you can place jar under `ZEPPELIN_HOME/interpreter/<interpreter_name>` directory if you're deploying jar without publishing to a package manager like maven.

I think the Helium application is not very well adopted among Zeppelin users. Any feedback (improvement point, potential use case, etc, etc) would be really appreciated.
